multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakEnvironmentInitalizationFilter.java

```diff
@@ -37,135 +36,146 @@
+    return claims.getSubject();
+  }
 
-      try {
-        String username =
-            claims.get(
-                keycloakSettings.get().get(KeycloakConstants.USERNAME_CLAIM_SETTING), String.class);
-        if (username == null) { // fallback to unique id promised by spec
-          // https://openid.net/specs/openid-connect-basic-1_0.html#ClaimStability
-          username = claims.getIssuer() + ":" + claims.getSubject();
-        }
-        String email = claims.get("email", String.class);
-        String id = claims.getSubject();
-
-        if (isNullOrEmpty(email)) {
-          boolean userNotFound = false;
-          try {
-            userManager.getById(id);
-          } catch (NotFoundException e) {
-            userNotFound = true;
-          }
-          if (userNotFound) {
-            try {
-              EnvironmentContext.getCurrent()
-                  .setSubject(new SubjectImpl(username, id, token, true));
-              Map<String, String> profileAttributes =
-                  keycloakProfileRetriever.retrieveKeycloakAttributes();
-              email = profileAttributes.get("email");
-              if (email == null) {
-                sendError(
-                    response,
-                    400,
-                    "Unable to authenticate user because email address is not set in keycloak profile");
-                return;
-              }
-            } finally {
-              EnvironmentContext.reset();
-            }
-          }
-        }
+  @Override
+  public Subject extractSubject(String token) throws ServletException {
 
-        User user = userManager.getOrCreateUser(id, email, username);
-        subject =
-            new AuthorizedSubject(
-                new SubjectImpl(user.getName(), user.getId(), token, false), permissionChecker);
-        session.setAttribute("che_subject", subject);
-      } catch (ServerException | ConflictException e) {
-        throw new ServletException(
-            "Unable to identify user " + claims.getSubject() + " in Che database", e);
-      }
-    }
+    Jws<Claims> jwt = jwtParser.parseClaimsJws(token);
+    Claims claims = jwt.getBody();
+    LOG.debug("JWT = {}", jwt);
+    // OK, we can trust this JWT
 
     try {
-      EnvironmentContext.getCurrent().setSubject(subject);
-      Span activeSpan = tracer.activeSpan();
-      if (activeSpan != null) {
-        TracingTags.USER_ID.set(tracer.activeSpan(), subject.getUserId());
+      String username =
+          claims.get(
+              keycloakSettings.get().get(KeycloakConstants.USERNAME_CLAIM_SETTING), String.class);
+      if (username == null) { // fallback to unique id promised by spec
+        // https://openid.net/specs/openid-connect-basic-1_0.html#ClaimStability
+        username = claims.getIssuer() + ":" + claims.getSubject();
+      }
+      String id = claims.getSubject();
+
+      String email = retrieveEmail(token, claims, username, id);
```

Comment:
if you use `Optional<String>` as a return value of `retrieveEmail`, you could have `String email = retrieveEmail(token, claims, username, id).orElseThrow(() -> new JwtException(...));`
I like it as it is more explicit that return value might be null. But it's personal preference, feel free to ignore if you don't like it.

Reply:
Makes sense